lib/aio/job.py

```diff
@@ -0,0 +1,191 @@
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import asyncio
+import contextlib
+import json
+import logging
+import platform
+import tempfile
+import textwrap
+import traceback
+from pathlib import Path
+from typing import Never
+
+from .abc import Forge, Subject
+from .jobcontext import JobContext
+from .jsonutil import JsonObject, get_dict, get_int, get_str, get_str_map, get_strv
+from .s3streamer import Index, LogStreamer
+from .spawn import run, spawn
+from .util import gather_and_cancel, read_utf8
+
+logger = logging.getLogger(__name__)
+
+BOTS_DIR = Path(__file__).parent.parent.parent
+
+
+class Failure(Exception):
+    pass
+
+
+class Job:
+    def __init__(self, obj: JsonObject) -> None:
+        # test subject specification
+        self.repo = get_str(obj, 'repo')
+        self.sha = get_str(obj, 'sha', None)
+        self.pull = get_int(obj, 'pull', None)
+        self.branch = get_str(obj, 'branch', None)
+        self.target = get_str(obj, 'target', None)
+
+        # test specification
+        self.container = get_str(obj, 'container', None)
+        self.secrets = get_strv(obj, 'secrets', ())
+        self.command = get_strv(obj, 'command', None)
+        self.env = get_str_map(obj, 'env')
+        self.timeout = get_int(obj, 'timeout', 120)
+
+        # reporting
+        self.context = get_str(obj, 'context')
+        self.slug = get_str(obj, 'slug', None)
+        self.title = get_str(obj, 'title', None)
+        self.report = get_dict(obj, 'report', None)
+
+
+async def timeout_minutes(minutes: float) -> Never:
+    await asyncio.sleep(60 * minutes)
+    raise Failure(f'Timeout after {minutes} minutes')
+
+
+async def poll_pr(api: Forge, repo: str, pull_nr: int, expected_sha: str) -> Never:
+    while True:
+        if reason := await api.check_pr_changed(repo, pull_nr, expected_sha):
+            raise Failure(reason)
+        await asyncio.sleep(60)
+
+
+async def run_container(job: Job, subject: Subject, ctx: JobContext, log: LogStreamer) -> None:
+    with tempfile.TemporaryDirectory() as tmpdir_path:
+        tmpdir = Path(tmpdir_path)
+        cidfile = tmpdir / 'cidfile'
+        attachments = tmpdir / 'attachments'
+
+        args = [
+            *ctx.container_cmd, 'run',
+            *ctx.container_run_args,
+            f'--cidfile={cidfile}',
+            '--interactive',
+            *(f'--env={k}={v}' for k, v in job.env.items()),
+            '--env=TEST_ATTACHMENTS=/tmp/attachments',
+            f'--env=COCKPIT_CI_LOG_URL={log.url}',
+            *(arg for name, arg in ctx.secrets_args.items() if name in job.secrets),
+            job.container or ctx.default_image,
+            'python3', '-c',
+            (BOTS_DIR / 'checkout-and-run').read_text(),  # lulz
+            f'--revision={subject.sha}',
```

Comment:
I was looking to avoid the usual selinux issues, and without having to relabel.

I'd be okay copying the script to the already existing temporary directory and letting the container relabel the copy?

Reply:
Ah, SELinux -- if it's easy and nice with a temp dir copy, do that, otherwise please just close this conversation. Not important.

Reply:
I did the copy.  It's working OK.